### PR TITLE
wasm: synchronise behaviour with non-strict-builtin-error mode in topdown

### PR DESCRIPTION
**Pull request by srenatus**
This allows us to fix most of the exceptions brought up in https://github.com/open-policy-agent/opa/issues/2954.

**However**, I've come across some WASM results **not** corresponding to non-strict topdown. The recipe for finding the discrepancies is:

1. change `exported_test.go` (topdown tests) to use `WithStrictBuiltinErrors(tc.StrictError)`
2. for every FAIL in the topdown tests, add `strict_error: true` to it's yaml test case file
3. see the test return an error (not undefined) in WASM evaluation.

An example of this is `2 % 0`: In topdown, it'll either throw `eval_builtin_error` or be `undefined` (non-strict), in WASM, it'll return an error.

⏩ Those errors have been addressed now. The evaluation logic of the "test/cases" tests is as follows:

- topdown: evaluates in strict mode if `tc.StrictError` is true, and checks the returned errors
- wasm: if `tc.StrictError` is true (i.e. the error depends on running in strict mode, which WASM doesn't), it asserts that there's no error, and the result set is empty.

Note that there's one case, the `"baseandvirtualdocs/enumerate virtual errors"`, where the turning OFF strict mode does not result in an empty result set, but in a partial one (the erroneous rules just don't add to the result). This is incidentally dealt with by checking the result (if provided in the test case) before (and in a separate conditional branch) checking for any errors.

----
Together with #3017, the only remaining exceptions are the ones related to jsonpatch ("stack frame moved" issue 💥) and the conflicting else keywords.

### CODE REVIEW

**Comment by srenatus** [comment]
Looking into further error strict/non-strict discrepancies.

**Comment by srenatus** [comment]
So....

- [x] added `strict_error: true` to those errors that only happen with _strict builtin errors_ enabled
- [x] adapted the wasm code where necessary to satisfy the non-strict-builtin-errors semantics
- [x] made the golang SDK ignore non-halt builtin errors accordingly

The remaining (partly new) exceptions are:
```
"baseandvirtualdocs/base/virtual: conflicts": "document merge conflict - https://github.com/open-policy-agent/opa/issues/2926"
"jsonpatch/set": "unexpected panic or evaluation error - https://github.com/open-policy-agent/opa/issues/2949"
"jsonpatch/json_patch_tests": "unexpected panic or evaluation error - https://github.com/open-policy-agent/opa/issues/2949"
"elsekeyword/conflict-2": "expected error missing - https://github.com/open-policy-agent/opa/issues/2954"
"baseandvirtualdocs/enumerate virtual errors": "TODO: update test case to deal with result"
```

☝️ the JSON and elsekeywords are tracked elsewhere; the first, `base/virtual: conflicts`, is dealt with in #3017.
The last TODO one I'll take care of in this PR.

**Comment by srenatus** [comment]
`"baseandvirtualdocs/enumerate virtual errors": "TODO: update test case to deal with result"` is done, this could use a review.

File: test/cases/testdata/jsonremove/test-jsonremove-0253.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 cases:
-- data: {}
+- data:
```
Comment (tsandall) [inline test/cases/testdata/jsonremove/test-jsonremove-0253.yaml:2]: 👀 learned something new about YAML today.

File: internal/wasm/sdk/test/e2e/external_test.go [hunk 2]
```diff
@@ -111,15 +111,22 @@ func assert(t *testing.T, tc cases.TestCase, result *opa.Result, err error) {
...
 		}
 		assertResultSet(t, *tc.WantResult, tc.SortBindings, result)
 	} else if tc.WantErrorCode != nil || tc.WantError != nil {
+		if tc.StrictError {
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			assertEmptyResultSet(t, result)
+			return
+		}
```
Comment (tsandall) [inline internal/wasm/sdk/test/e2e/external_test.go:129]: It took me a minute to understand what's going on here--a short comment could be helpful. Something like "the wasm compiler does not support strict errors so if the error condition is only visible when strict errors are enabled, expect an empty/undefined result from evaluation."

**Review by tsandall — DISMISSED** [review]
LGTM. Just one comment on the wasm test runner change--mostly made to clarify my own understanding. Thanks for updating all of the existing tests!

File: test/cases/testdata/jsonremove/test-jsonremove-0253.yaml (hunk 1 above)
Comment (srenatus) [reply test/cases/testdata/jsonremove/test-jsonremove-0253.yaml:2]: yeah it also doesn't matter 😄

File: internal/wasm/sdk/test/e2e/external_test.go (hunk 2 above)
Comment (srenatus) [reply internal/wasm/sdk/test/e2e/external_test.go:129]: Thanks, I like that sentence. I'll use it. 😎

**Comment by srenatus** [comment]
Rebased, and fixed the comment. I'll squash and merge when green.

**Pull request merged**